src/Scripts/UI/Controls/GraphicsEditor.gd

```diff
@@ -135,17 +149,65 @@ func setup_from_image(image_: Image):
 
 	image = image_
 
+	# Ensure the undo history contains the initial state
+	layer_undo_histories[_draw_layer.name] = []
```

Comment:
use stack functionality for storing the undo actions on the graphics editor (godot does not have stack data structure but the same functionality can be achieved using arrays)